org.opentosca.bus.management.service.impl/src/org/opentosca/bus/management/service/impl/util/ParameterHandler.java

```diff
@@ -0,0 +1,298 @@
+                    }
+                }
+            }
+        } else {
+            ParameterHandler.LOG.warn("Unable to update input parameters with RelationshipTemplateInstance equal to null!");
+        }
+
+        return inputParams;
+    }
+
+    /**
+     * Adds an entry to the given map if it does not already contain an entry with the same key.
+     *
+     * @param map the map to update
+     * @param key the key of the entry to update
+     * @param value the value which is set if there is not yet an entry with the key
+     */
+    private static void putOnlyIfNotSet(final Map<String, String> map, final String key, final String value) {
+        if (!map.containsKey(key)) {
+            map.put(key, value);
+        }
+    }
+
+    /**
+     * Returns the input parameters of the given operation which are specified in the TOSCA
+     * definitions of the NodeType or RelationshipType.
+     *
+     * @param csarID ID of the CSAR which contains the NodeType or RelationshipType with the
+     *        operation
+     * @param typeID ID of the NodeType or RelationshipType which contains the operation
+     * @param interfaceName the name of the interface which contains the operation
+     * @param operationName the operation name for which the parameters are searched
+     *
+     * @return specified input parameters of the operation
+     */
+    private static List<String> getExpectedInputParams(final CSARID csarID, final QName typeID,
+                                                       final String interfaceName, final String operationName) {
+
+        ParameterHandler.LOG.debug("Fetching expected input params of " + operationName + " in interface "
+            + interfaceName);
+        final List<String> inputParams = new ArrayList<>();
+
+        ParameterHandler.LOG.debug("Checking for params with Type: {}", typeID);
+        if (ServiceHandler.toscaEngineService.hasOperationOfATypeSpecifiedInputParams(csarID, typeID, interfaceName,
+                                                                                      operationName)) {
+
+            final Node definedInputParameters =
+                ServiceHandler.toscaEngineService.getInputParametersOfATypeOperation(csarID, typeID, interfaceName,
+                                                                                     operationName);
+
+            if (definedInputParameters != null) {
+
+                final NodeList definedInputParameterList = definedInputParameters.getChildNodes();
+
+                for (int i = 0; i < definedInputParameterList.getLength(); i++) {
+
+                    final Node currentNode = definedInputParameterList.item(i);
+
+                    if (currentNode.getNodeType() == Node.ELEMENT_NODE) {
+
+                        final String name = ((Element) currentNode).getAttribute("name");
+
+                        inputParams.add(name);
+                    }
+                }
+            }
+        }
```

Comment:
arrow code. When fixing, consider to replace early returns with `return Collections.emptyList()` to avoid the extra List allocation